src/community.c

```diff
@@ -3567,3 +3567,1094 @@ int igraph_i_compare_communities_rand(const igraph_vector_t *v1,
+ * next iteration. At each iteration all clusters are guaranteed to be
+ * connected and well-separated. After an iteration in which nothing has
+ * changed, all nodes and some parts are guaranteed to be locally optimally
+ * assigned. Finally, asymptotically, all subsets of all clusters are
+ * guaranteed to be locally optimally assigned. For more details, please see
+ * Traag, Waltman &amp; van Eck (2019).
+ *
+ * The objective function being optimized is
+ *
+ * 1 / 2m sum_ij (A_ij - gamma n_i n_j)d(s_i, s_j)
+ *
+ * where m is the total edge weight, A_ij is the weight of edge (i, j), gamma is
+ * the so-called resolution parameter, n_i is the node weight of node i, s_i is
+ * the cluster of node i and d(x, y) = 1 if and only if x = y and 0 otherwise.
+ * By setting n_i = k_i, the degree of node i, and dividing gamma by 2m, you
+ * effectively obtain an expression for modularity. Hence, the standard
+ * modularity will be optimized when you supply the degrees as \c node_weights
+ * and by supplying as a resolution parameter 1.0/(2*m), with m the number of
+ * edges.
+ *
+ * \param graph The input graph. It must be an undirected graph.
+ * \param edge_weights Numeric vector containing edge weights. If \c NULL, every edge
+ *    has equal weight of 1. The weights need not be non-negative.
+ * \param node_weights Numeric vector containing node weights.
+ * \param resolution_parameter The resolution parameter used, which is
+ *    represented by gamma in the objective function mentioned in the
+ *    documentation.
+ * \param beta The randomness used in the refinement step when merging. A small
+ *    amount of randomness (beta = 0.01) typically works well.
+ * \param membership The membership vector. This is both used as the initial
+ *    membership from which optimisation starts and is updated in place. It
+ *    must hence be properly initialized. When finding clusters from scratch it
+ *    is typically started using a singleton clustering. This can be achieved
+ *    using \c igraph_vector_init_seq.
+ * \param nb_clusters The number of clusters contained in \c membership. Must
+ *    not be a \c NULL pointer.
+ * \param quality The quality of the partition, in terms of the objective
+ *    function as included in the documentation. If \c NULL the quality will
+ *    not be calculated.
+ * \return Error code.
+ *
+ * Time complexity: near linear on sparse graphs.
+ *
+ * \example examples/simple/igraph_community_leiden.c
+ */
+int igraph_community_leiden(const igraph_t *graph,
+  const igraph_vector_t *edge_weights, const igraph_vector_t *node_weights,
+  const igraph_real_t resolution_parameter, const igraph_real_t beta,
+  igraph_vector_t *membership, igraph_integer_t *nb_clusters, igraph_real_t *quality) 
+{
+  igraph_vector_t *i_edge_weights, *i_node_weights;
+  int ret;
+
+  if (igraph_vector_size(membership) != igraph_vcount(graph)) 
+    IGRAPH_ERROR("Membership length does not equal the number of vertices", IGRAPH_EINVAL);
```

Comment:
Other community detection functions only require the membership vector to be initialized, but not to have a certain size. It will be resized automatically to the correct size.

Reply:
Yes, that is true, but this function starts optimisation from a given membership vector, which will be updated inplace. It would be possible to check for a membership vector of an incorrect length and then initialize it with a singleton partition. However, that may only lead to confusion if people pass a membership vector of the correct length, without intending to. This is why I opted for this solution.

Reply:
E.g. `igraph_community_leading_eigenvector` has a separate boolean `start` argument, which controls whether the user-supplied `membership` vector should be used as a starting point, or whether the membership vector should be initialized automatically.

Reply:
Here's how it does the checks: https://github.com/igraph/igraph/blob/master/src/community.c#L1591

Reply:
OK, good point. I will also use that.

Reply:
N.B. `nb_clusters` should also be checked or set correctly (in the case `start=TRUE`).

Reply:
`nb_clusters` does not need to be checked, it is set when cleaning the membership.